src/c_kzg_4844.c

```diff
@@ -1098,34 +1100,46 @@ static C_KZG_RET compute_challenges(BLSFieldElement *out, BLSFieldElement r_powe
   const size_t np = ni + n * BYTES_PER_BLOB;
   const size_t nb = np + n * 48;
 
-  uint8_t* bytes = calloc(nb + (n == 0), sizeof(uint8_t)); // need at least 1 byte more than ni for hash later
+  uint8_t* bytes = calloc(nb, sizeof(uint8_t));
   if (bytes == NULL) return C_KZG_MALLOC;
 
+  /* Copy domain seperator */
   memcpy(bytes, FIAT_SHAMIR_PROTOCOL_DOMAIN, 16);
   bytes_of_uint64(&bytes[16], FIELD_ELEMENTS_PER_BLOB);
   bytes_of_uint64(&bytes[16 + 8], n);
 
+  /* Copy polynomials */
   for (i = 0; i < n; i++)
     for (j = 0; j < FIELD_ELEMENTS_PER_BLOB; j++)
       bytes_from_bls_field(&bytes[ni + BYTES_PER_FIELD_ELEMENT * (i * FIELD_ELEMENTS_PER_BLOB + j)], &polys[i][j]);
 
+  /* Copy commitments */
   for (i = 0; i < n; i++)
     bytes_from_g1(&bytes[np + i * 48], &comms[i]);
 
-  uint8_t hash_output[32] = {0};
-  hash(hash_output, bytes, nb);
-  memcpy(bytes, hash_output, 32);
-  bytes[32] = 0x0;
-  hash(hash_output, bytes, 33);
-
-  if (n > 0) {
-    if (n > 1) bytes_to_bls_field(&r_powers[1], hash_output);
-    compute_powers(r_powers, n);
-  }
-
-  bytes[32] = 0x1;
-  hash(hash_output, bytes, 33);
-  bytes_to_bls_field(out, hash_output);
+  /* Now let's create challenges! */
+  uint8_t hashed_data[32] = {0};
+  hash(hashed_data, bytes, nb);
+
+  /* Compute r */
+  uint8_t r_bytes[32] = {0};
```

Comment:
Would at prefer to allocate only one of `r_bytes` and `eval_challenge`, and possibly neither.

Reply:
Similar to the case above, this is largely a matter of readability. I thought that these two extra variables make the code look closer to the spec. If you think they are confusing, we can remove them!

Reply:
I could go either way on these, but would find it clearer with reusing `hashed_data` for all hash outputs. (It could be called `hash_output` if the other one is called `hash_input`.)

Reply:
Since you could go either way I went naughty and didn't change this part of the code. I think this way it's more verbose but the variable names stay close to the spirit of the spec. IMO that will be better for people who review/audit the code.